File: Medium/22. Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> res;
        string s;
        dfs(res, s, n, n);
        return res;
    }

    void dfs(vector<string> &res, string s, int left, int right)
    {
        if (left == 0 && right == 0)
        {
            res.push_back(s);
            return;
        }
        if (left > 0)
        {
            dfs(res, s + "(", left - 1, right);
        }
        if (right > left)
        {
            dfs(res, s + ")", left, right - 1);
        }
    }
};

File: Medium/22. Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        if n == 0:
            return []
        res = []
        self.dfs(n, n, "", res)
        return res

    def dfs(self, left, right, path, res):
        if left == 0 and right == 0:
            res.append(path)
            return
        if left > 0:
            self.dfs(left - 1, right, path + "(", res)
        if right > left:
            self.dfs(left, right - 1, path + ")", res)
